lib/puppet/util/autoload.rb

```diff
@@ -204,20 +207,20 @@ def load(name, env = nil)
   #
   # This uses require, rather than load, so that already-loaded files don't get
   # reloaded unnecessarily.
-  def loadall(env = nil)
+  def loadall(env = Puppet.lookup(:current_environment))
```

Comment:
@pcarlisle @hlindberg It seems there is a common pattern among 3x functions to explicitly call `loadall` to ensure a different function they want to call has been loaded, e.g. https://github.com/example42/puppet-tp/blob/master/lib/puppet/parser/functions/tp_content.rb#L16. Same is true for `loaded?`, e.g. https://github.com/puppetlabs/puppetlabs-stdlib/blob/4.12.x/lib/puppet/parser/functions/has_ip_network.rb#L17-L18.

So making `environment` be a required parameter would break a non-trivial amount of modules.

So one idea:

1. Make the `Puppet::Util::Autoload` public methods require an environment parameter in 6. Raise an error if environment is nil.
2. For 3x functions, create a `FunctionAutoloader` subclass, which defaults the current environment and passes that to the superclass.
3. Have the `FunctionAutoloader` subclass emit a deprecation warning in 6.x , when `loadall` or `load` or `loaded?` is called, instructing people to just call the 3x function directly, e.g. as described in [self.send(:myfunction)](https://tickets.puppetlabs.com/browse/PUP-4058?focusedCommentId=145112&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-145112)

Reply:
Need to think about the implications.
On (3.) they should use `Scope#call_function('myfunction', ...)` as that works for both 4.x and 3.x functions (self.send(:function_foo) does not work with 4.x).

Reply:
Instead of writing a "FunctionAutoLoader" we could write a 3xFunctionLoader for 4.x loaders (I wrote one a long time ago but we never got around to actually using it because of the 4x loaders restriction on visibility into other modules based on dependencies). Solving that would also mean that we can make progress on relaxing that. It would help with the performance issue if the search is 4x function intertwined with 3x as opposed to first checking all 4.x and then 3.x.